### PR TITLE
[DO NOT MERGE] Add support for Vector Mass and Diffusion using libCEED.

### PR DESCRIPTION
**Pull request by YohannDudouit**
| PR | Author | Editor | Reviewers | Assignment | Approval | Merge| 
 | --- | --- | --- | --- | --- | --- | --- | 
| [#1713](https://github.com/mfem/mfem/pull/1713) | @YohannDudouit | @tzanio | @camierjs + @v-dobrev + @pazner | 08/23/20 | ⌛due 09/06/20 | ⌛due 10/07/20 | |

### CODE REVIEW

File: fem/bilininteg.hpp
Comment (barker29) [inline fem/bilininteg.hpp:2198]: You need a destructor here with `delete ceedDataPtr` inside `#ifdef MFEM_USE_CEED` guards, as is done in `MassIntegrator` and `DiffusionIntegrator`.

File: fem/bilininteg.hpp
Comment (barker29) [inline fem/bilininteg.hpp:2593]: Same comment as before.

File: fem/bilininteg.hpp
Comment (YohannDudouit) [reply fem/bilininteg.hpp:2198]: Thank you, I added the destructor.

File: fem/bilininteg.hpp
Comment (YohannDudouit) [reply fem/bilininteg.hpp:2600]: done.

File: fem/bilininteg_vecdiffusion.cpp
Comment (camierjs) [inline fem/bilininteg_vecdiffusion.cpp:520]: There is a lot of places with `#ifdef MFEM_USE_CEED` followed by `if (DeviceCanUseCeed())`: couldn't we find a way (perhaps not in this PR) to avoid or generalize this 'difference' and `#ifdef` treatment?
Like `libCEED`, it would be good to be able to fuse the 3 kernels.

File: fem/libceed/ceed.cpp
Comment (pazner) [inline fem/libceed/ceed.cpp:405]: The QFunctionContext interface needs to be updated here too (see #1701 and #1705).

File: fem/bilininteg.hpp
Comment (pazner) [inline fem/bilininteg.hpp:2142]: Some of these `#ifdef` blocks could be removed if `CeedData*` is defined as `void*` when `MFEM_USE_CEED` is not defined. It might make the code a little bit less noisy.

File: fem/bilininteg_vecdiffusion.cpp
Comment (pazner) [inline fem/bilininteg_vecdiffusion.cpp:136]: The `if` statement is not needed.

File: fem/bilininteg_vecdiffusion.cpp
Comment (pazner) [inline fem/bilininteg_vecdiffusion.cpp:140]:
```suggestion
      ceedDataPtr = new CeedData;
      InitCeedCoeff(Q, ceedDataPtr);
      return CeedPADiffusionAssemble(fes, *ir, * ceedDataPtr);
```

File: fem/bilininteg_vecmass.cpp
Comment (pazner) [inline fem/bilininteg_vecmass.cpp:41]:
```suggestion
      delete ceedDataPtr;
      ceedDataPtr = new CeedData;
      InitCeedCoeff(Q, ceedDataPtr);
      return CeedPAMassAssemble(fes, *ir, *ceedDataPtr);
```

Same changes also apply in `bilininteg_mass_pa.cpp` and `bilininteg_diffusion_pa.cpp`.

**Comment by YohannDudouit** [comment]
@camierjs @pazner You seem to both want to change the way the `CeedData` is stored, I can indeed remove the type inside each `Integrator` and use `static_cast` when I want to use it. However, for the destructor I will need to cast there also, but I will need to `#ifdef` guard the cast because the type is not defined, I could `delete` a `void*` but this doesn't call the destructor and would not destroy the libCEED objects properly.
I have another proposition, I could put the `#ifdef` inside the `CeedData` type, so that the type is always defined, I could also encapsulate the other `#ifdef` inside some `libCEED` functions, I think I could factorize all the `#ifdef` inside `ceed.hpp` by doing this. What do you think about this?

**Comment by pazner** [comment]
> @camierjs @pazner You seem to both want to change the way the `CeedData` is stored, I can indeed remove the type inside each `Integrator` and use `static_cast` when I want to use it. However, for the destructor I will need to cast there also, but I will need to `#ifdef` guard the cast because the type is not defined, I could `delete` a `void*` but this doesn't call the destructor and would not destroy the libCEED objects properly.
> I have another proposition, I could put the `#ifdef` inside the `CeedData` type, so that the type is always defined, I could also encapsulate the other `#ifdef` inside some `libCEED` functions, I think I could factorize all the `#ifdef` inside `ceed.hpp` by doing this. What do you think about this?

I think your second proposal is similar to what I was suggesting. In `ceed.hpp`, if `MFEM_USE_CEED` is not defined, `CeedData` can be defined to be e.g. an empty struct. That way, the `CeedData *` pointers in the integrators do not need to be `#ifdef`'d. There shouldn't be any need for `static_cast` or trouble with `delete` as far as I can see.

**Comment by YohannDudouit** [comment]
@camierjs @pazner I moved the `#ifdef MFEM_USE_CEED` mostly in in `ceed.hpp` and ceed.cpp` tell me if you like it.

**Comment by pazner** [comment]
I think it looks good! How can I test it out?

**Comment by YohannDudouit** [comment]
@pazner I think the easiest is to modify `ex1` to use `VectorMass` and `VectorDiffusion`, I will be adding tests to #1719 too.

**Comment by barker29** [comment]
@pazner you could also look at @jandrej 's [branch](https://github.com/mfem/mfem/tree/navier-ceed) where he uses these in the Navier mini-app for the definite Helmholtz operator.

**Comment by pazner** [comment]
This will segfault at exit because of the same issue in the destructor for `CeedData` as mentioned here: https://github.com/mfem/mfem/pull/1670#pullrequestreview-479931734

**Comment by YohannDudouit** [comment]
@pazner isn't the destructor bug already fixed in master?

**Comment by YohannDudouit** [comment]
To facilitate merging all the libCEED PR into master, I can merge them into #1719 once they are approved, like this we can merge only #1719 into next and master, does that sound better?

**Comment by tzanio** [comment]
> To facilitate merging all the libCEED PR into master, I can merge them into #1719 once they are approved, like this we can merge only #1719 into next and master, does that sound better?

I haven't looked at all the PRs involved, but if they are connected + are relatively simple, fewer PRs is better

File: fem/libceed/ceed.hpp
Comment (pazner) [inline fem/libceed/ceed.hpp:95]: Also `CeedQFunctionContext build_ctx` needs to be destroyed or there will be a memory leak

Comment (YohannDudouit) [reply fem/libceed/ceed.hpp:96]: done.

File: fem/bilininteg.hpp
Comment (pazner) [inline fem/bilininteg.hpp:2117]: Can you put `Q_order(0)` before `Q(&q)` here? The compiler will complain about initialization order otherwise (`nvcc` doesn't complain, but with `MFEM_USE_CUDA=NO` you should see it).

Comment (YohannDudouit) [reply fem/bilininteg.hpp]: done.

**Comment by YohannDudouit** [comment]
This branch is merged and tested in #1719.

**Comment by YohannDudouit** [comment]
@camierjs @v-dobrev @pazner  Once you've reviewed and approved this PR, could you please check that the merge in #1719 looks good to you, and if it does write "The merge of #1713 in #1719 looks good to me."
I will close this PR afterwards in favor of #1719 to avoid complex merge conflicts for the editors.

File: fem/libceed/ceed.cpp
Comment (v-dobrev) [inline fem/libceed/ceed.cpp:581]: Why is this disabled?

File: fem/libceed/ceed.cpp
Comment (v-dobrev) [inline fem/libceed/ceed.cpp:98]: Possible optimization: the ceed backend may need to copy the data from host to device -- we can avoid that copy if the mesh nodes, `*mesh->GetNodes()`, are already available in device memory.

File: fem/libceed/ceed.cpp
Comment (v-dobrev) [inline fem/libceed/ceed.cpp:154]: Possible optimization: avoid host-to-device copy by using the device pointer from `*ceedCoeff->coeff`.

File: fem/libceed/ceed.cpp
Comment (v-dobrev) [inline fem/libceed/ceed.cpp:207]: Instead of `Device::Allows(Backend::CUDA)`, it is probably better to use `Device::Allows(Backend::DEVICE_MASK)` in preparation for other device backends like "ceed-hip".

File: fem/libceed/ceed.cpp
Comment (v-dobrev) [inline fem/libceed/ceed.cpp:239]: Similar to above: `Backend::CUDA` --> `Backend::DEVICE_MASK`.

File: fem/libceed/ceed.cpp
Comment (YohannDudouit) [reply fem/libceed/ceed.cpp:581]: Using the address of the pointer as a hash key creates bugs, @jeremylt is looking into a more robust key.

File: fem/libceed/ceed.cpp
Comment (YohannDudouit) [reply fem/libceed/ceed.cpp:207]: I prefer to do it when we'll actually add the feature, I think it's premature.

File: fem/libceed/ceed.cpp
Comment (YohannDudouit) [reply fem/libceed/ceed.cpp:154]: This optimization is already done in `test-ceed` is it okay if I leave this one like this here? I also moved the `CeedVector` definition into `InitCeedCoeff`.

Comment (YohannDudouit) [reply fem/libceed/ceed.cpp:165]:
```c++
   else if (GridFunctionCoefficient* coeff =
               dynamic_cast<GridFunctionCoefficient*>(Q))
   {
      CeedGridCoeff *ceedCoeff = new CeedGridCoeff;
      ceedCoeff->coeff = coeff->GetGridFunction();
      CeedVectorCreate(internal::ceed, ceedCoeff->coeff->FESpace()->GetNDofs(),
                       &ceedCoeff->coeffVector);
      CeedScalar *d_ptr;
      CeedMemType mem;
      CeedGetPreferredMemType(internal::ceed, &mem);
      if ( Device::Allows(Backend::CUDA) && mem==CEED_MEM_DEVICE )
      {
         d_ptr = const_cast<CeedScalar*>(ceedCoeff->coeff->Read());
      }
      else
      {
         d_ptr = const_cast<CeedScalar*>(ceedCoeff->coeff->HostRead());
         mem = CEED_MEM_HOST;
      }
      CeedVectorSetArray(ceedCoeff->coeffVector, mem, CEED_USE_POINTER,
                         d_ptr);
      ptr->coeff_type = CeedCoeff::Grid;
      ptr->coeff = static_cast<void*>(ceedCoeff);
   }
```

File: fem/libceed/ceed.cpp
Comment (YohannDudouit) [reply fem/libceed/ceed.cpp]: done.

File: fem/libceed/ceed.cpp
Comment (YohannDudouit) [reply fem/libceed/ceed.cpp]: done.

File: fem/bilininteg.hpp
Comment (YohannDudouit) [reply fem/bilininteg.hpp]: I think I have addressed this, can you check if you like it?

File: fem/libceed/ceed.cpp
Comment (YohannDudouit) [reply fem/libceed/ceed.cpp]: done.

File: fem/libceed/ceed.cpp
Comment (YohannDudouit) [reply fem/libceed/ceed.cpp:592]: Or we need to find a way to remove the entries in the map when we destroy a `FiniteElementSpace`.

Comment (YohannDudouit) [reply fem/libceed/ceed.cpp]: Okay, I pushed a fix, tell me if you like it.

Comment (YohannDudouit) [reply fem/libceed/ceed.cpp]: I added a `RemoveCeedBasisAndRestriction` function that is called in the destructor of `FiniteElementSpace`.

Comment (v-dobrev) [reply fem/libceed/ceed.cpp]: This will be a bigger change (maybe for another PR) but I think we can get rid of the global containers and instead associate the ceed restrictions with the `FiniteElementSpace` objects and the ceed bases with the `FiniteElement` objects. This is what we do for the mfem versions of the restriction and "basis" (aka `DofToQuad`) objects.

Comment (YohannDudouit) [reply fem/libceed/ceed.cpp]: I like the idea, I think @jeremylt had something like this in mind. I can do it in an other PR.

Comment (jeremylt) [reply fem/libceed/ceed.cpp]: Does a FE object hold both a FES and a IR?

This is the sort of information I was talking about on slack that would be nice to have in some developer documentation.

File: fem/libceed/ceed.cpp
Comment (v-dobrev) [reply fem/libceed/ceed.cpp:165]: Okay. You can also update `Backend::CUDA` to `Backend::DEVICE_MASK` there.

In the future, this way of reviewing and merging PRs should definitely be avoided. The burden of tracking the changes from this PR into `test-ceed` (#1719) is asking too much from the reviewers.

If there are big conflicts, then some of the PRs will wait for the conflicting ones to be merged in `master` first, so that the author can resolve the conflicts in his PR before merging into `next`.

File: fem/libceed/ceed.cpp
Comment (v-dobrev) [inline fem/libceed/ceed.cpp:109]: This logic for wrapping an `mfem::Vector` as `CeedVector` is used in a number of places, so we should probably turn these into functions to reduce duplication.

Comment (YohannDudouit) [reply fem/libceed/ceed.cpp:109]: This is a good idea.

Comment (YohannDudouit) [reply fem/libceed/ceed.cpp:109]: I've done it directly in #1719 , there was even more code duplication there.

File: fem/libceed/ceed.cpp
Comment (v-dobrev) [reply fem/libceed/ceed.cpp]: @jeremylt, here are some notes on that.

The `FiniteElement` objects do not know about `FiniteElementSpace` and they does not own any `IntegrationRule` objects.

`FiniteElementSpace` objects have a pointer to a `FiniteElementColloection`, which they do not own, and which can be queried for `FiniteElement` objects, e.g. based on the element types present in the mesh. Also, `FiniteElementSpace` has a pointer to a `Mesh` which it does not own.

`FiniteElement` objects can be asked for a "basis" description (aka `DofToQuad`) for a given quadrature rule using: https://github.com/mfem/mfem/blob/a40523a7cfe6bb0037311cd46200e2a5eddfc180/fem/fe.hpp#L561-L565

The returned `DofToQuad` object is lazy-constructed, cached, and owned by the `FiniteElement`.

Restriction operators are similarly handled by `FiniteElementSpace` objects using the method: https://github.com/mfem/mfem/blob/a40523a7cfe6bb0037311cd46200e2a5eddfc180/fem/fespace.hpp#L337-L354

**Comment by YohannDudouit** [comment]
@v-dobrev Do you think I should add a `ceed` namespace below the `mfem` one?

File: fem/libceed/mass.cpp
Comment (YohannDudouit) [inline fem/libceed/mass.cpp:28]: If you know the formula for the other cases, I can implement them in an other PR.

File: fem/libceed/diffusion.cpp
Comment (YohannDudouit) [inline fem/libceed/diffusion.cpp:28]: @v-dobrev If you know the formula for the other cases, I can implement them in an other PR.

**Comment by v-dobrev** [comment]
> @v-dobrev Do you think I should add a `ceed` namespace below the `mfem` one?

Right now, I think all ceed related classes/types/functions in the `mfem` namespace have `Ceed` in their name, so they should not conflict with other definitions. Putting them in a the `mfem::ceed` namespace is an alternative but I don't see a big difference. If you think it is worth changing, it is better to do it in a separate PR after #1719 is merged.

File: fem/libceed/mass.cpp
Comment (v-dobrev) [reply fem/libceed/mass.cpp:28]: No need to do it right now. This was just to inform/protect users who may try to use this case.

File: fem/libceed/diffusion.cpp
Comment (v-dobrev) [reply fem/libceed/diffusion.cpp:28]: Same here, no need to do it right now.

**Comment by v-dobrev** [comment]
LGTM

**Review by v-dobrev — APPROVED** [review]

File: fem/libceed/mass.cpp
Comment (v-dobrev) [reply fem/libceed/mass.cpp:28]: For completeness: in the general case, `det(J)` is replaced by `[det(J^T J)]^{1/2}`, where `J` is the `SpaceDimension` x `Dimension` Jacobian matrix.

File: fem/libceed/diffusion.cpp
Comment (v-dobrev) [reply fem/libceed/diffusion.cpp:28]: For completeness: if `J` is the `SpaceDimension` x `Dimension` Jacobian matrix, then gradients are mapped from reference to physical space using the matrix `J (J^T J)^{-1}`, so the expression for a quadrature point becomes
```
[det(J^T J)]^{1/2} (J^T J)^{-1} J^T Q J (J^T J)^{-1}
```
where `Q` is the diffusivity coefficient matrix. Substituting `(J^T J)^{-1} = adj(J^T J)/det(J^T J)`, the expression can be simplified as
```
[det(J^T J)]^{-3/2} X^T Q X,   where X = J adj(J^T J)
```
When `J` is square, `J (J^T J)^{-1} = J^{-T} = adj(J)^T/det(J)` and the simplified form is
```
[det(J)]^{-1} Y^T Q Y,   where Y = adj(J)^T
```

File: fem/libceed/ceed.cpp
Comment (pazner) [inline fem/libceed/ceed.cpp:288]: Does it make sense to use `auto &entry : ` here instead of `auto entry : `?

Comment (YohannDudouit) [reply fem/libceed/ceed.cpp]: I don't think it makes a big difference in this case, but it doesn't hurt.

Comment (pazner) [reply fem/libceed/ceed.cpp]: Yeah I think the copy should be very cheap anyway

**Review by pazner — APPROVED** [review]

File: fem/libceed/ceed.cpp
Comment (YohannDudouit) [reply fem/libceed/ceed.cpp]: I wish we could use decomposition declaration:
```c++
for (auto &[key, value] : internal::ceed_basis_map)
```

**Review by camierjs — APPROVED** [review]
Looks good!

**Comment by tzanio** [comment]
Please close this after it was merged in #1719.

**Pull request closed without merging**
